### PR TITLE
nixos/network-interfaces: remove network-interfaces.target

### PR DESCRIPTION
**Pull request by nyabinary**
## Description of changes
Seemingly nothing is using this and network-interfaces was deprecated all the way back in 2016: https://github.com/NixOS/nixpkgs/pull/18491

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by pbsds** [comment]
This will be a silent failure for [many](https://github.com/search?type=code&q=-repo%3Anixos%2Fnixpkgs+lang%3Anix+network-interfaces.target) people.

Can/will this be caught at eval time?

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by SuperSandro2000** [comment]
Then definitely add a release note entry.

I don't think it is as big. Some of the results are stone age old nixpkgs copies.

When we add `-is:fork` then we are already down to just below 200 https://github.com/search?type=code&q=-repo%3Anixos%2Fnixpkgs+lang%3Anix+network-interfaces.target+-is%3Afork

**Review by Mindavi — APPROVED** [review]

**Review by NickCao — APPROVED** [review]

**Review by pbsds — APPROVED** [review]

File: nixos/doc/manual/release-notes/rl-2405.section.md
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2405.section.md:30]:
```suggestion
- The legacy and long deprecated systemd target `network-interfaces.target` has been removed.
```

Comment (Nyabinary) [reply nixos/doc/manual/release-notes/rl-2405.section.md]: Done :3

**Review by lilyinstarlight — APPROVED** [review]
This is a good idea, thank you!

I looked through [some of the downstream repos](https://github.com/search?type=code&q=-repo%3Anixos%2Fnixpkgs+lang%3Anix+network-interfaces.target+-path%3Anixos%2Fmodules+-path%3Amodules%2Fservices+-path%3Amodules%2Ftasks+-path%3Amodules%2Fsystem+-is%3Afork) and I see a few that are still active (including nixops depending on this unit...), so maybe we should make a short post at https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574 to notify people?

**Review by philiptaron — APPROVED** [review]

**Comment by SuperSandro2000** [comment]
> Why not add a warning or an assert failure for now and remove the whole code in 24.11?

That would be here https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/systemd.nix#L431-L452

**Comment by nyabinary** [comment]
> > Why not add a warning or an assert failure for now and remove the whole code in 24.11?
> 
> That would be here https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/systemd.nix#L431-L452

So is it ready to merge after I rebase?

**Review by mkg20001 — APPROVED** [review]

**Comment by mkg20001** [comment]
> > > Why not add a warning or an assert failure for now and remove the whole code in 24.11?
> > 
> > 
> > That would be here [`master`/nixos/modules/system/boot/systemd.nix#L431-L452](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/systemd.nix?rgh-link-date=2023-12-20T16%3A40%3A12Z#L431-L452)
> 
> So is it ready to merge after I rebase?

You would have to add an assertion in that particular file, then it should be good.

**Review by philiptaron — APPROVED** [review]
What's needed in order to merge this?

**Comment by mkg20001** [comment]
> You would have to add an assertion in that particular place, then it should be good.

Added one

**Comment by nyabinary** [comment]
> > You would have to add an assertion in that particular place, then it should be good.
> 
> Added one

Thank you :)

File: nixos/doc/manual/release-notes/rl-2405.section.md
Comment (mweinelt) [inline nixos/doc/manual/release-notes/rl-2405.section.md:44]:
```suggestion
- The legacy and long deprecated systemd target `network-interfaces.target` has been removed. Use the `network.target` instead.
```

I think since it was basically just wrapping that, and `network.target` has those exact semantics (network interfaces are now available) this is a sound recommendation.

**Comment by mkg20001** [comment]
Created pr https://github.com/NixOS/nixpkgs/pull/276493 for warning on 23.11 release

File: nixos/doc/manual/release-notes/rl-2405.section.md
Comment (nyabinary) [reply nixos/doc/manual/release-notes/rl-2405.section.md]: It should just be ``Use `network.target` instead.`` omitting “the” from the sentence, right?

**Comment by mkg20001** [comment]
I'd need a review on the assertions. If it's good or a bit too hacky what I wrote. Then I'll merge.

**Comment by nyabinary** [comment]
@mkg20001 

It should just be use `network.target` instead for the release notes. Omitting the “the” from the sentence, right?

File: nixos/modules/system/boot/systemd.nix
Comment (philiptaron) [inline nixos/modules/system/boot/systemd.nix:461]: I think this is a fine set of things to check. It could be mentioned in a bunch of other places, such as conflicts or before, and likely others.

Frankly this assertion, though requested, feels like gold-plating. If it helps this review go through, fine.

**Review by philiptaron — APPROVED** [review]

File: nixos/modules/system/boot/systemd.nix
Comment (mkg20001) [reply nixos/modules/system/boot/systemd.nix:461]: The most common usages were in those parameters. I didn't find a sane way to just check all the places where it could occur.

File: nixos/modules/system/boot/systemd.nix
Comment (ElvishJerricco) [inline nixos/modules/system/boot/systemd.nix:460]: This assertion is very confusingly written. Why is there a `concatLists` for a list with a single element? Why is it mapping a list made with `optional` to an assertion with `assertion = false` instead of just putting the condition in the `assertion` field?

Comment (mkg20001) [reply nixos/modules/system/boot/systemd.nix:460]: > assertion = false

I don't know the impact of adding a lot of assertions (as there would be one for every system service)

> concatLists

This is copied from the warnings above, to support multiple assertions later.

Comment (nyabinary) [reply nixos/modules/system/boot/systemd.nix:460]: Any further objections? @ElvishJerricco

**Comment by nyabinary** [comment]
Fixed merge conflict should be ready to merge I think(?)

**Comment by mkg20001** [comment]
@nyabinary You force-pushed away the assertion, re-added that.

**Comment by nyabinary** [comment]
> @nyabinary You force-pushed away the assertion, re-added that.

Oops my bad sorry lol

File: nixos/modules/system/boot/systemd.nix
Comment (ElvishJerricco) [reply nixos/modules/system/boot/systemd.nix:460]: I still think this was extremely weirdly written and should have been cleaned up, but I wasn't going to block the PR on it

**Comment by RaitoBezarius** [comment]
If we can believe this stuff: https://github.com/NixOS/nixpkgs/runs/20435679228 — this had visibly no impact on evaluation performance, phew, module system is great!

**Comment by SuperSandro2000** [comment]
That's probably because the expensive part, evaluating things, is already done in warnings and you only see a difference if both would be dropped. According to testing a year ago evaluating warnings takes 4s on a noop rebuild.

**Pull request merged**
